KEP-2799: Reduction of Secret-based Service Account Tokens (#2800)
kubernetes/enhancements: keps/sig-auth/2799-reduction-of-secret-based-service-account-token/README.md

[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes actions to reduce the surface area of secret-based service
account tokens.

## Motivation

As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
would be obtained via TokenRequest API and stored as projected volume. This
change obviates the need for auto-generation of secret-based service account
tokens which are [less secure than the bound token](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background).

### Goals

- No auto-generation of secret-based service account token.
- Removal of unused auto-generated secret-based service account tokens

### Non-Goals

## Proposal

- In release 1.23, change the service account control loop in Token Controller
  to not create secret for the service account. At the same time, warn usage of
  auto-created secret-based service account tokens and enrouage users to use
  TokenRequest API or manually-created secret-based service account tokens.

> - In release 1.25, purge auto-generated secret-based service account tokens.

Author: done

### Notes/Constraints/Caveats

- A warning mechanism should be implemented to help users
  migrate.
- Only clean up auto-generated tokens which:
  - are not referenced by pods
  - have not been used to authenticate for some duration (time duration or number of releases)
- To consult active usage of secret-based tokens, metric
  `serviceaccount_legacy_tokens_total` or audit annotation
  `authentication.k8s.io/legacy-token` could be used.

### Risks and Mitigations

- In release 1.24 (LegacyServiceAccountTokenNoAutoGeneration Beta), consumers
  depending directly on waiting for and reading tokens out of auto-generated
  secrets might stop working. To mitigate,
  1. In 1.23, announce the stopping auto-generating token secrets will happen in
     1.24
  2. Publish pointers to TokenRequest or the manual secret request flow.
- In release 1.25 (LegacyServiceAccountTokenCleanUp Beta), usage of secret-based
  token that are referenced in service accounts might stop working. To mitigate,

>   add poiners of TokenRequest API and manually created tokens in the validation

Author: done

  result.

## Design Details

### Token Controller

#### LegacyServiceAccountTokenNoAutoGeneration:

1. Token Controller stops to create secret for service accounts.

> 2. To falicitate LegacyServiceAccountTokenCleanUp, when this feature

Author: done

   is Beta in 1.24 where pod-based usage of auto-generated tokens that are injected
   by admission controller should have dried up,
   - Annotate auto-generated secrets with a date indicating when we started
     tracking use (e.g."secret.kubernetes.io/used-since: \$date", default to "never").
   - When a legacy token is used, if it is from a secret with the `used-since`
     annotation, issue a warning, update the value, and record in a metric.
     Optionally, add a label `in-use` for fast query.

#### LegacyServiceAccountTokenCleanUp

Token Controller starts to remove unused auto-generated secrets (secrets
bi-directionally referenced by the service account) and not mounted by pods.

In release 1.25 when this feature is Beta and enabled by default, delete secrets
iff they have `used-since` label and the time duration since the date is over

> a sufficient period of time that can be configured by cluster admins.

Reviewer: what do we intend the default to be here?

Author: 4 months = 1 release? we neither have metric to know if this is a good value nor any constrain that is against this value.

### Test Plan

- Unit tests
- E2E tests

> - Upgrade tests

Author: done

- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

#### LegacyServiceAccountTokenCleanUp

| Alpha | Beta | GA   |
| ----- | ---- | ---- |
| 1.24  | 1.25 | 1.26 |

#### Beta -> GA Graduation

- [ ] In use by multiple distributions
- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

#### Alpha -> Beta Graduation

- [ ] In use by multiple distributions
- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

### Upgrade / Downgrade Strategy

TODO

### Version Skew Strategy

The only touches control plane, so version skew strategy is not applicable.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: LegacyServiceAccountTokenNoAutoGeneration
  - Components depending on the feature gate: kube-controller-manager
  - Feature gate name: LegacyServiceAccountTokenCleanUp:
  - Components depending on the feature gate: kube-controller-manager

###### Does enabling the feature change any default behavior?

> - LegacyServiceAccountTokenNoAutoGeneration: no.

Author: done

-->

###### Does this feature depend on any specific services running in the cluster?

<!--
Think about both cluster-level services (e.g. metrics-server) as well
as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

> no.

Author: done

###### Will enabling / using this feature result in introducing new API types?

no.

###### Will enabling / using this feature result in any new calls to the cloud provider?

no.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

> no.

Author: done

  depending directly on waiting for and reading tokens out of auto-generated
  secrets might stop working. To mitigate,
  1. When Alpha, announce the stopping auto-generating token secrets will happen
     in Beta.
  2. Emit warnings when using auto-generated token secrets.
  3. Publish pointers to TokenRequest or the manual secret request flow.
- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
  secret-based token might stop working. To mitigate,
  1. When Alpha, annouce the cleanup starts at Beta
  2. Emit warnings when using auto-generated token secrets.
  3. Add pointers of TokenRequest API and manually created tokens in the validation
     result.

## Design Details

### Token Controller

#### LegacyServiceAccountTokenNoAutoGeneration:

1. Token Controller stops to create secret for service accounts.
2. To facilitate LegacyServiceAccountTokenCleanUp, when this feature
   is Alpha where pod-based usage of auto-generated tokens that are injected
   by admission controller should have dried up (BoundServiceAccountTokenVolume
   is GAed in 1.22),
   - Annotate auto-generated secrets with a date indicating when we started

>      tracking use (e.g."secret.kubernetes.io/used-since: \$date", default to "unknown").

Author: > it clears last-used (if set) on secrets when they are used

we don't need to do this. this could be problematic as well in case envelop encryption is used. when determining the time that a secret was used:
1. `last-used` if exists and after `tracked-since`
2. defaults to `tracked-since`

we can s/LegacyServiceAccountTokenCleanUp/LegacyServiceAccountTokenNoAutoGeneration to fast forward.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes actions to reduce the surface area of secret-based service
account tokens.

## Motivation

As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
would be obtained via TokenRequest API and stored as projected volume. This
change obviates the need for auto-generation of secret-based service account
tokens which are [less secure than the bound token](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background).

### Goals

- No auto-generation of secret-based service account token.
- Removal of unused auto-generated secret-based service account tokens

### Non-Goals

- Removal of [explicitly requested secret-based service account tokens](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token).

## Proposal

> - Change the service account control loop in Token Controller to not create

Reviewer: ```suggestion
- Change the service account control loop in Token Controller to not auto-create
```

Author: done

- Auto generated secret-based service account tokens are those requested by
  Token Controller.
- Only clean up auto-generated tokens which:
  - are not referenced by pods
  - have not been used to authenticate for some duration (time duration or number of releases)
- To consult active usage of secret-based tokens, metric
  `serviceaccount_legacy_tokens_total` or audit annotation
  `authentication.k8s.io/legacy-token` could be used.

### Risks and Mitigations

- When feature LegacyServiceAccountTokenNoAutoGeneration is Beta, consumers
  depending directly on waiting for and reading tokens out of auto-generated
  secrets might stop working. To mitigate,
  1. Emit warnings when using auto-generated token secrets.
  2. Publish pointers to TokenRequest or the manual secret request flow.
- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
  secret-based token might stop working. To mitigate,
  1. When Alpha, annouce the cleanup starts at Beta
  2. Emit warnings when using auto-generated token secrets.
  3. Add pointers of TokenRequest API and manually created tokens in the validation
     result.

## Design Details

### LegacyServiceAccountTokenNoAutoGeneration:

> Token Controller stops to create secret for service accounts. This feature would

Reviewer: ```suggestion
Token Controller stops auto-creating secrets for service accounts. This feature would
```

Author: done

2. defaults to `tracked-since`

### Test Plan

- Unit tests
- Integration tests
  - Previously auto-generated secret-based token that's used within the
    configurable cleanup duration will continue to work.
  - Previously auto-generated secret-based token that's used after the
    configurable cleanup duration will be deleted.
- E2E tests
  - Secret-based tokens would not be auto-generated.
  - Still able to explicitly request a secret-based token.
  - The explicitly requested token would not be deleted.

### Graduation Criteria

#### LegacyServiceAccountTokenNoAutoGeneration

| Alpha | Beta | GA   |
| ----- | ---- | ---- |
| -     | 1.24 | 1.25 |

#### Beta -> GA Graduation

- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

#### Alpha -> Beta Graduation

- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing
- [ ] Document and communicate the available actions that consumers of
      auto-generated secret-based tokens should take. (migrate to either use
      tokenrequest or explicitly request secret-based tokens)

#### LegacyServiceAccountTokenTracking

| Alpha | Beta | GA   |
| ----- | ---- | ---- |

> | -     | 1.24 | 1.25 |

Author: makes sense.

- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
  secret-based token might stop working. To mitigate,
  1. When Alpha, annouce the cleanup starts at Beta
  2. Emit warnings when using auto-generated token secrets.
  3. Add pointers of TokenRequest API and manually created tokens in the validation
     result.

## Design Details

### LegacyServiceAccountTokenNoAutoGeneration:

Token Controller stops auto-creating secret for service accounts. This feature would
be enabled when it is implemented since no new code is added and this can make
sure new clusters are in good state.

### LegacyServiceAccountTokenTracking

To facilitate LegacyServiceAccountTokenCleanUp, we implement a simple controller
in kube-apiserver that maintains a bool value configmap in `kube-system` to
indicates if tracking is enabled in the cluster. It is similar to the existing
`ClusterAuthenticationTrustController` that maintains `configmap/extension-apiserver-authentication`.

- When LegacyServiceAccountTokenTracking is enabled in all apiservers,

  - the controller creates/updates a configmap in `kube-system` namespace that
    stores the current date as `tracked-since`.

>   - when a legacy token is used, issue a warning, annotate/update the

Reviewer 2: you'll annotate legacy tokens on every usage?

Reviewer: plan was to do so at date granularity... max of one write per 24 hour period

Author: captured in the KEP.

    in a periodic way.

### LegacyServiceAccountTokenCleanUp

Token Controller starts to remove unused auto-generated secrets (secrets
bi-directionally referenced by the service account) and not mounted by pods.

When this feature is Beta and enabled by default, delete secrets iff it is over
a sufficient period of time (one year by default) since last used. The period
can be configured by cluster admins.

Determine the date that a given secret was last used:

1. `last-used` if exists and after `tracked-since`.
2. defaults to `tracked-since`

### Test Plan

- Unit tests
- Integration tests
  - Previously auto-generated secret-based token that's used within the
    configurable cleanup duration will continue to work.
  - Previously auto-generated secret-based token that's used after the
    configurable cleanup duration will be deleted.
- E2E tests
  - Secret-based tokens would not be auto-generated.
  - Still able to explicitly request a secret-based token.
  - The explicitly requested token would not be deleted.

### Graduation Criteria

#### LegacyServiceAccountTokenNoAutoGeneration

| Alpha | Beta | GA   |
| ----- | ---- | ---- |

> | -     | 1.24 | 1.25 |

Reviewer 2: Can you describe why this starts as beta?

Author: because in 1.24, all pods should be admitted in 1.22+ and they should be using bound tokens. this feature gate mostly guard deletions of code.

Reviewer 2: > because in 1.24, all pods should be admitted in 1.22+ and they should be using bound tokens. this feature gate mostly guard deletions of code.

I meant in the KEP itself. It's unusual to start featuregates in beta.  Not unheard of, but worthy of explaining in the doc why it's more reasonable to start this one at beta than at alpha.  Why should is it more reasonable to start this as opt-out than as opt-in?

Author: updated.

### Version Skew Strategy

The only touches control plane, so version skew strategy is not applicable.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: LegacyServiceAccountTokenNoAutoGeneration
  - Components depending on the feature gate: kube-controller-manager
  - Feature gate name: LegacyServiceAccountTokenTracking
  - Components depending on the feature gate: kube-apiserver
  - Feature gate name: LegacyServiceAccountTokenCleanUp:
  - Components depending on the feature gate: kube-controller-manager

###### Does enabling the feature change any default behavior?

- LegacyServiceAccountTokenNoAutoGeneration: no legacy tokens are auto-generated.
- LegacyServiceAccountTokenTracking: legacy tokens would have new annotation and a configmap would be created in kube-system.
- LegacyServiceAccountTokenCleanUp: unused auto-generated legacy tokens will be removed.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> yes for all feature gates.

Reviewer 2: do all eight combinations of these featuregates make sense?

Author: each feature gate is independent from each other. any combination would continue to work.

###### What happens if we reenable the feature if it was previously rolled back?

> - LegacyServiceAccountTokenNoAutoGeneration: the same as enable the feature.

Reviewer 2: clarify that this controller will create tokens for serviceaccounts created while the feature was off.

Author: done.

  - [Optional] Aggregation method:
  - Components exposing the metric: kube-apiserver

LegacyServiceAccountTokenNoAutoGeneration and LegacyServiceAccountTokenCleanUp
might cause few workloads to fail but there is no way for us to inject metric
in workloads to detect this.

###### What are the reasonable SLOs (Service Level Objectives) for the above SLIs?

none. we expect the number recorded in the above metric going down in the long
term.

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

none.

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

no.

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

> up to one additional write request per day could be made to auto-generated secrets still in use.

Reviewer 2: I don't see where in the KEP above this "per day" comes from.  Where is the frequency determined?

Author: when the legacy token is used, we check the annotation and if it was annotated at the same day, we don't write again. i updated the KEP above

Reviewer 2: > when the legacy token is used, we check the annotation and if it was annotated at the same day, we don't write again. i updated the KEP above

thanks

  - have not been used to authenticate for some duration (time duration or number of releases)
- To consult active usage of secret-based tokens, metric
  `serviceaccount_legacy_tokens_total` or audit annotation
  `authentication.k8s.io/legacy-token` could be used.

### Risks and Mitigations

- When feature LegacyServiceAccountTokenNoAutoGeneration is Beta, consumers
  depending directly on waiting for and reading tokens out of auto-generated
  secrets might stop working. To mitigate,
  1. Emit warnings when using auto-generated token secrets.
  2. Publish pointers to TokenRequest or the manual secret request flow.
- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
  secret-based token might stop working. To mitigate,
  1. When Alpha, annouce the cleanup starts at Beta
  2. Emit warnings when using auto-generated token secrets.
  3. Add pointers of TokenRequest API and manually created tokens in the validation
     result.

## Design Details

### LegacyServiceAccountTokenNoAutoGeneration:

Token Controller stops to create secret for service accounts. This feature would
be enabled when it is implemented since no new code is added and this can make

> sure new clusters are in good state.

Reviewer: Populate the `LegacyServiceAccountTokenNoAutoGeneration` table below proposing beta in 1.24. That also means the PRR section needs beta-level completion for this aspect.

### LegacyServiceAccountTokenTracking

To facilitate LegacyServiceAccountTokenCleanUp,

> - When LegacyServiceAccountTokenTracking is disabled,

Reviewer: ```suggestion
- When LegacyServiceAccountTokenTracking is enabled,
```

  - create a configmap in `kube-system` namespace that stores the current

>     date as `tracked-since` if it doesn't exist yet.

Reviewer: clarify it's the API server that does this (prior art for API server persisting data to a kube-system configmap in pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go)

also clarify when this happens (could either be done once at apiserver startup or periodically... I'd probably suggest periodically)

  - when a legacy token is used, issue a warning, annotate/update the
    `last-used` on the secret, and record in a metric. optionally, add a
    label `in-use` for fast query.

- When LegacyServiceAccountTokenTracking is disabled,

>   - delete the configmap in `kube-system` namespace.

Reviewer: clarify it's the API server that does this, and whether this happens at server start, or periodically, or in response to add/update events on the configmap

consider these scenarios:
* avoid a misleading tracked-since configmap being persisted when a tracking-enabled and tracking-disabled API server are running concurrently
* ensure that once the last tracking-disabled API server leaves the cluster, a tracked-since configmap will eventually be created without the tracking-enabled API servers needing to be restarted

### LegacyServiceAccountTokenCleanUp

Token Controller starts to remove unused auto-generated secrets (secrets
bi-directionally referenced by the service account) and not mounted by pods.

When this feature is Beta and enabled by default, delete secrets iff it is over
a sufficient period of time (one year by default) since last used. The period
can be configured by cluster admins.

> Determin the date that a given secret was last used:

Reviewer: ```suggestion
Determine the date that a given secret was last used:
```

2. defaults to `tracked-since`

### Test Plan

- Unit tests
- Integration tests
  - Previously auto-generated secret-based token that's used within the
    configurable cleanup duration will continue to work.
  - Previously auto-generated secret-based token that's used after the
    configurable cleanup duration will be deleted.
- E2E tests
  - Secret-based tokens would not be auto-generated.
  - Still able to explicitly request a secret-based token.
  - The explicitly requested token would not be deleted.

### Graduation Criteria

#### LegacyServiceAccountTokenNoAutoGeneration

| Alpha | Beta | GA  |
| ----- | ---- | --- |
| -     | -    | -   |

#### Beta -> GA Graduation

- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

#### Alpha -> Beta Graduation

- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing
- [ ] Document and communicate the available actions that consumers of
      auto-generated secret-based tokens should take. (migrate to either use
      tokenrequest or explicitly request secret-based tokens)

#### LegacyServiceAccountTokenTracking

| Alpha | Beta | GA  |
| ----- | ---- | --- |

> | -     | -    | -   |

Reviewer: fill proposed timeline here?

#### Beta -> GA Graduation

- [ ] In use by multiple distributions
- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

#### Alpha -> Beta Graduation

- [ ] In use by multiple distributions
- [ ] Approved by PRR and scalability
- [ ] Any known bugs fixed
- [ ] Tests passing

### Upgrade / Downgrade Strategy

#### LegacyServiceAccountTokenCleanUp

| Alpha | Beta | GA  |
| ----- | ---- | --- |

> | -     | -    | -   |

Reviewer: fill proposed timeline here?

sure new clusters are in good state.

### LegacyServiceAccountTokenTracking

To facilitate LegacyServiceAccountTokenCleanUp, we implement a simple controller
in kube-apiserver that maintains a bool value configmap in `kube-system` to
indicates if tracking is enabled in the cluster. It is similar to the existing
`ClusterAuthenticationTrustController` that maintains `configmap/extension-apiserver-authentication`.

- When LegacyServiceAccountTokenTracking is enabled in all apiservers,

  - the controller creates/updates a configmap in `kube-system` namespace that
    stores the current date as `tracked-since`.
  - when a legacy token is used, issue a warning, annotate/update the
    `last-used` on the secret at date granularity, and record in a metric.
    optionally, add a label `in-use` for fast query.

- When LegacyServiceAccountTokenTracking is disabled in any apiserver,
  - the controller ensures the configmap in `kube-system` namespace is deleted
    in a periodic way.

### LegacyServiceAccountTokenCleanUp

Token Controller starts to remove unused auto-generated secrets (secrets
bi-directionally referenced by the service account) and not mounted by pods.

> When this feature is Beta and enabled by default, delete secrets iff it is over

Reviewer 2: Is the delete contingent on the "tracked-since" configmap?  You have to be sure that the last-used seen on service-account secrets is still valid somehow, otherwise disabling the LegacyServiceAccountTokenTracking featuregate and enabling LegacyServiceAccountTokenCleanUp becomes a combination that doesn't make sense to allow.

Author: yes, `tracked-since` is required in determining when the secret is lastly used.
